source/server/active_tcp_listener.cc

```diff
@@ -16,51 +16,30 @@
 namespace Envoy {
 namespace Server {
 
-namespace {
-void emitLogs(Network::ListenerConfig& config, StreamInfo::StreamInfo& stream_info) {
-  stream_info.onRequestComplete();
-  for (const auto& access_log : config.accessLogs()) {
-    access_log->log(nullptr, nullptr, nullptr, stream_info);
-  }
-}
-} // namespace
-
 ActiveTcpListener::ActiveTcpListener(Network::TcpConnectionHandler& parent,
                                      Network::ListenerConfig& config)
-    : ActiveTcpListener(
-          parent,
+    : TypedActiveStreamListenerBase<ActiveTcpConnection>(
+          parent, parent.dispatcher(),
           parent.dispatcher().createListener(config.listenSocketFactory().getListenSocket(), *this,
                                              config.bindToPort(), config.tcpBacklogSize()),
-          config) {}
+          config),
+      tcp_conn_handler_(parent) {
+  config.connectionBalancer().registerHandler(*this);
+}
 
 ActiveTcpListener::ActiveTcpListener(Network::TcpConnectionHandler& parent,
                                      Network::ListenerPtr&& listener,
                                      Network::ListenerConfig& config)
-    : ActiveListenerImplBase(parent, &config), parent_(parent), listener_(std::move(listener)),
-      listener_filters_timeout_(config.listenerFiltersTimeout()),
-      continue_on_listener_filters_timeout_(config.continueOnListenerFiltersTimeout()) {
+    : TypedActiveStreamListenerBase<ActiveTcpConnection>(parent, parent.dispatcher(),
+                                                         std::move(listener), config),
+      tcp_conn_handler_(parent) {
   config.connectionBalancer().registerHandler(*this);
 }
 
 ActiveTcpListener::~ActiveTcpListener() {
-  is_deleting_ = true;
   config_->connectionBalancer().unregisterHandler(*this);
 
-  // Purge sockets that have not progressed to connections. This should only happen when
-  // a listener filter stops iteration and never resumes.
-  while (!sockets_.empty()) {
-    ActiveTcpSocketPtr removed = sockets_.front()->removeFromList(sockets_);
-    parent_.dispatcher().deferredDelete(std::move(removed));
-  }
-
-  for (auto& chain_and_connections : connections_by_context_) {
-    ASSERT(chain_and_connections.second != nullptr);
-    auto& connections = chain_and_connections.second->connections_;
-    while (!connections.empty()) {
-      connections.front()->connection_->close(Network::ConnectionCloseType::NoFlush);
-    }
-  }
-  parent_.dispatcher().clearDeferredDeleteList();
+  cleanupConnections();
```

Comment:
Could this work be done by the base class destructor?

Reply:
I would suggest undoing this part for now since it is less of a mechanical refactor, and doing it in a separate PR so it can get the attention it deserves.  I'm worried about some of these changes making the class hierarchy significantly harder to follow.  deferredRemoveFilterChains and the use of the curiously recurring template pattern is another example.

Reply:
Also see: https://github.com/envoyproxy/envoy/pull/16947#discussion_r652275477

Reply:
removed as per recent comment